lib/Transforms/IPO/StructAnalysisCloseProximity.cpp

```diff
@@ -0,0 +1,1004 @@
+// nodes in its subtree into a table stored with node. With collapsing, a tree
+// can be reduced to a node with a table, as long as the CP relations between
+// nodes in the table is updated. Thus a FRG will eventually be reduced to a
+// root nodes with a relation table, or a much smaller tree with a few
+// non-collapsed nodes. In either case, the complexity is much smaller than
+// exponential that can complete with reasonable amount of time.
+//
+//===------------------------------------------------------------------------===//
+
+#include "StructFieldCacheAnalysisImpl.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/Analysis/CFG.h"
+#include "llvm/Support/Format.h"
+
+#include <cmath>
+#include <queue>
+
+using namespace llvm;
+
+#define DEBUG_PRINT_COUNT(x) (format("%.2f", (x)))
+#define DEBUG_PRINT_DIST(x) (format("%.3f", (x)))
+
+static cl::opt<bool> PerformFRGOnly(
+    "struct-analysis-FRG-only", cl::init(false), cl::Hidden,
+    cl::desc("Stop the analysis after performing FRG generation"));
+
+static cl::opt<bool> PerformCPGCheck(
+    "struct-analysis-check-CPG", cl::init(false), cl::Hidden,
+    cl::desc("Perform CPG checking algorithm that takes a long time"));
+
+static cl::opt<bool> DisableIgnoreZeroCountNodes(
+    "struct-analysis-disable-ignore-zeros", cl::init(false), cl::Hidden,
+    cl::desc("Ignore zero-count nodes in FRG generation"));
+
+// Utility functions
+// Function used to update (count, distance) pair by taking a weighted average
+// of the Result and Source and save result to Result. Used when updating
+// CPG or updating an edge weight
+static void updatePairByMerging(ExecutionCountType &ResultCount,
+                                DataBytesType &ResultDistance,
+                                ExecutionCountType SourceCount,
+                                DataBytesType SourceDistance) {
+  if (SourceCount + ResultCount != 0)
+    ResultDistance =
+        (SourceCount * SourceDistance + ResultCount * ResultDistance) /
+        (SourceCount + ResultCount);
+  else
+    ResultDistance = 0;
+  ResultCount += SourceCount;
+}
+
+// Function used to update (count, distance) pair. distance is updated by
+// directly adding ResultDistance and SourceDistance. count is updated by
+// taking minimum of a portion of ResultCount and a portion of SourceCount
+// Used when calculating edge weight along a path, especially when nodes
+// have multiple in and out edges
+static void updatePairByConnecting(ExecutionCountType &ResultCount,
+                                   double OutRatio,
+                                   DataBytesType &ResultDistance,
+                                   ExecutionCountType SourceCount,
+                                   double InRatio,
+                                   DataBytesType SourceDistance) {
+  assert(InRatio <= 1 && OutRatio <= 1);
+  ResultCount = std::min(ResultCount * OutRatio, SourceCount * InRatio);
```

Comment:
I don't quite understand this comment clearly. Can you elaborate more? Do you want me to use two functions to update them separately? Or you don't think it's a good idea to use reference?